BIP32: Add Deserialize extended key target

kuliq23 opened the pull request:
### NEW TARGET: Deserialization of a BIP32 extended key
This target accepts a Bip32 extended key in the base58check-encoded form (string of b58 chars), deserializes it, then returns it as a string (return may change to bool, or different parts of the key).
In order to make it even remotely possible to send a valid extkey, formatting of the random bytes is needed.

TLDR bip32 ExtKey serialization from [https://en.bitcoin.it/wiki/BIP_0032](https://en.bitcoin.it/wiki/BIP_0032)
> Extended public and private keys are serialized as follows:

> * 4 byte: version bytes (mainnet: 0x0488B21E public, 0x0488ADE4 private; testnet: 0x043587CF public, 0x04358394 private)
> * 1 byte: depth: 0x00 for master nodes, 0x01 for level-1 derived keys, ....
> * 4 bytes: the fingerprint of the parent's key (0x00000000 if master key)
> * 4 bytes: child number. This is ser32(i) for i in xi = xpar/i, with xi the key being serialized. (0x00000000 if master key)
> * 32 bytes: the chain code
> * 33 bytes: the public key or private key data (serP(K) for public keys, 0x00 || ser256(k) for private keys)

> This 78 byte structure can be encoded like other Bitcoin data in Base58, by first adding 32 checksum bits (derived from the double SHA-256 checksum), and then converting to the Base58 representation. This results in a Base58-encoded string of up to 112 characters. Because of the choice of the version bytes, the Base58 representation will start with "xprv" or  "xpub" on mainnet, "tprv" or "tpub" on testnet.

My idea was to create multiple custom mutators, each "helping the target" a bit more, to maximize the range tested.
(e.g. Mut1 - encode random bytes to b58, Mut2 - hardcode the first 4 version bytes + Mut1, Mut3 - fill to 78 bytes + Mut2, Mut4 - append correct checksum + Mut3, ...)

### I see 2 issues with this approach:

#### 1. Base58 utils access
For these mutations, base58 functions are needed, so far i only added a placeholder mutator that creates seemingly b58 looking, 78B long data.  I see 3 solutions:
* Custom implementation in main.cpp (similarly to bitcoin_double_sha256() in main.cpp)
* "Borrow" it from Core "modules/bitcoin/base58.h"
* Add the base58 utils to "modules/custommutator/" (similarly to how bech32 and sha256 are there) <- optimal?

#### 2. How mutators work
Please correct me if i am wrong:

As mutators are selected in the building process, running the fuzzer means that only an explicit set of them can be active for that running instance at a time....
Does that mean that if we e.g. want to fuzz "long-term", multiple instances of bitcoinfuzz need to be running to cover all targets? (each with a specified mutator?)
Is this something to avoid (shall I reavaluate the multi-mutator solution and instead e.g. implement multiple targets that begin with mutating the input?


Thanks.

kuliq23 commented:
> 2. Decode base58 (remove checksum) -> raw 78 bytes

Does this mean that the initial `fuzz_data` that the mutator receives is a (valid) key encoded in base58? I thought the mutator receives random bytes and works with them... Is this to do with the "corpus"? (Does the corpus provide an initial set of correct values?)

> 5. Encode back to base58
> 6. Recalculate and add checksum (double SHA-256, first 4 bytes)

nit: I believe these steps should be switched to comply with the scheme.

erickcestari commented:
> Does this mean that the initial `fuzz_data` that the mutator receives is a (valid) key encoded in base58? I thought the mutator receives random bytes and works with them... Is this to do with the "corpus"? (Does the corpus provide an initial set of correct values?)

The `fuzz_data` depends on the fuzzer's state: initially when there's no corpus, the fuzzer generates random bytes, but once a corpus exists, it uses inputs from the corpus which are valid BIP32-formatted keys (base58-encoded with checksums, since our custom mutator ensures this format). In both cases, the mutator attempts to decode the input to raw bytes, and if decoding fails (e.g., for random initial bytes), we either return early or fall back to a minimal valid input (see the [code here](https://github.com/bitcoinfuzz/bitcoinfuzz/blob/aa89a7fbe41d21b1771b7c099f0c6d125b9d6af6/main.cpp#L373)). Once we have raw bytes, the mutator modifies them and then re-serializes the result into proper BIP32 format with a valid checksum, ensuring that even mutations of random data eventually produce structurally valid keys for the target function to process.

> 
> > 5. Encode back to base58
> > 6. Recalculate and add checksum (double SHA-256, first 4 bytes)
> 
> nit: I believe these steps should be switched to comply with the scheme.

Yes, true

kuliq23 committed:
feat: add bip32_deserialize_extended_key target for RUST_BITCOIN, BITCOIN_CORE, NBITCOIN

Edit/fix deserialize modules and mutator

modules no longer use ENV to get extkeytype pre-hand

first symbol of extkey used for testnet vs mainnet in btccore as setting the parameters is per process and changing them mid run causes issues

NBitcoin accepts keys of all 4 types that throw non-zero child number in master key error in rust_bitcoin and result in empty key-bytes in bitcoin-core

erickcestari commented on custommutator/mutators/extended_key.h:
```diff
@@ -0,0 +1,101 @@
+
+    // pad remaining bytes if needed
+    if (returned_size < BIP32_PAYLOAD_LEN) {
+        std::memset(scratch.data() + returned_size, 0, BIP32_PAYLOAD_LEN - returned_size);
```
After adding the pad to the scratch, the returned size should be set to BIP32_PAYLOAD_LEN.


```suggestion
    if (returned_size < BIP32_PAYLOAD_LEN) {
        std::memset(scratch.data() + returned_size, 0, BIP32_PAYLOAD_LEN - returned_size);
        returned_size = BIP32_PAYLOAD_LEN;
```

kuliq23 replied:
good catch!

erickcestari commented on custommutator/mutators/extended_key.h:
```diff
@@ -0,0 +1,101 @@
+#ifndef BITCOINFUZZ_CUSTOMMUTATOR_EXTENDED_KEY_H
```
Nice custom mutator! I've made a few changes, including setting the default to a random prefix instead of the xpub fallback.

```suggestion
#ifndef BITCOINFUZZ_CUSTOMMUTATOR_EXTENDED_KEY_H
#define BITCOINFUZZ_CUSTOMMUTATOR_EXTENDED_KEY_H

#include <custommutator/utils/base58.h>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <span>
#include <string>
#include <string_view>
#include <vector>
#include <array>

extern "C" size_t LLVMFuzzerMutate(uint8_t *Data, size_t Size, size_t MaxSize);

constexpr size_t VERSION_PREFIX_LEN = 4;
constexpr size_t BIP32_PAYLOAD_LEN = 78;

struct VersionBytes {
    uint8_t data[VERSION_PREFIX_LEN];
};

constexpr std::array<VersionBytes, 4> VERSION_PREFIXES = {{
    {{0x04, 0x88, 0xB2, 0x1E}},  // xpub
    {{0x04, 0x88, 0xAD, 0xE4}},  // xprv
    {{0x04, 0x35, 0x87, 0xCF}},  // tpub
    {{0x04, 0x35, 0x83, 0x94}},  // tprv
}};

constexpr size_t IDX_XPUB = 0;
constexpr size_t IDX_XPRV = 1;
constexpr size_t IDX_TPUB = 2;
constexpr size_t IDX_TPRV = 3;

const VersionBytes& GetVersionBytes(std::string_view key_type) {
    if (key_type == "xprv") return VERSION_PREFIXES[IDX_XPRV];
    if (key_type == "tpub") return VERSION_PREFIXES[IDX_TPUB];
    if (key_type == "tprv") return VERSION_PREFIXES[IDX_TPRV];
    return VERSION_PREFIXES[IDX_XPUB];
}

const VersionBytes& GetRandomVersion(unsigned int seed) {
    return VERSION_PREFIXES[seed % VERSION_PREFIXES.size()];
}

extern "C" size_t LLVMFuzzerCustomMutator(uint8_t *fuzz_data, size_t size, 
                                          size_t max_size, unsigned int seed) {
    
    // Decode input from base58check
    std::string_view input{reinterpret_cast<const char*>(fuzz_data), size};
    std::vector<unsigned char> payload(BIP32_PAYLOAD_LEN, 0x00);
    
    std::vector<unsigned char> decoded;
    if (DecodeBase58Check(std::string{input}, decoded, static_cast<int>(max_size))) {
        const size_t copy_len = std::min(decoded.size(), BIP32_PAYLOAD_LEN);
        std::memcpy(payload.data(), decoded.data(), copy_len);
    } else {
        const size_t copy_len = std::min(size, BIP32_PAYLOAD_LEN);
        std::memcpy(payload.data(), fuzz_data, copy_len);
    }
    
    // Mutate the payload
    size_t mutated_size = LLVMFuzzerMutate(payload.data(), BIP32_PAYLOAD_LEN, BIP32_PAYLOAD_LEN);
    
    // Ensure we have exactly BIP32_PAYLOAD_LEN bytes
    if (mutated_size < BIP32_PAYLOAD_LEN) {
        std::memset(payload.data() + mutated_size, 0, BIP32_PAYLOAD_LEN - mutated_size);
    }
    payload.resize(BIP32_PAYLOAD_LEN);
    
    // Apply version prefix: use env if set, otherwise random
    const char* env = std::getenv("EXTKEYTYPE");
    const auto& version = env ? GetVersionBytes(env) : GetRandomVersion(seed);
    std::memcpy(payload.data(), version.data, VERSION_PREFIX_LEN);
    
    // Encode back to base58check
    std::string encoded = EncodeBase58Check(std::span<const unsigned char>{payload});
    
    // Copy result to output buffer
    const size_t output_len = std::min(encoded.size(), max_size);
    std::memcpy(fuzz_data, encoded.data(), output_len);

    return output_len;
}

#endif //BITCOINFUZZ_CUSTOMMUTATOR_EXTENDED_KEY_H
```

kuliq23 replied:
thx

erickcestari commented on modules/rustbitcoin/rust_bitcoin_lib/src/lib.rs:
```diff
@@ -268,3 +273,74 @@ pub unsafe extern "C" fn rust_bitcoin_bip32_master_keygen(
+    };
+    //get the keytype from env variable, default is xpub
+    let extkeytype = env::var("EXTKEYTYPE").unwrap_or_else(|_| "xpub".to_string());
+    //xprv, tprv
+    if extkeytype == "xprv" || extkeytype == "tprv" {
+        //try decode xprv, tprv
+        if let Ok(xprv) = Xpriv::from_str(&ext_str) {
+            //format the result to string
+            let result = format_ext_key_common(
+                xprv.depth,
+                xprv.parent_fingerprint,
+                xprv.child_number,
+                xprv.chain_code,
+                &xprv.private_key.secret_bytes(), // as bytes
+            );
+            //return formatted string
+            str_to_c_string(&result)
+        } else {
+            str_to_c_string("INVALID")
+        }
+    //xpub, tpub
+    } else {
+        //try decode xpub, tpub
+        if let Ok(xpub) = Xpub::from_str(&ext_str) {
+            //format the result to string
+            let result = format_ext_key_common(
+                xpub.depth,
+                xpub.parent_fingerprint,
+                xpub.child_number,
+                xpub.chain_code,
+                &xpub.public_key.serialize(), // as bytes
+            );
+            //return formatted string
+            str_to_c_string(&result)
+        } else {
+            str_to_c_string("INVALID")
+        }
+    }
+
+}
```
We shouldn't retrieve the environment here. Instead, we should try decoding the input using the bip32 key types. We should also do this for the other modules.

```suggestion
#[no_mangle]
pub unsafe extern "C" fn rust_bitcoin_bip32_deserialize_extended_key(
    data: *const u8,
    len: usize,
) -> *mut c_char {
    let data_slice = slice::from_raw_parts(data, len);
    let ext_str = match std::str::from_utf8(data_slice) {
        Ok(s) => s,
        Err(_) => return str_to_c_string("INVALID"),
    };

    if let Ok(xprv) = Xpriv::from_str(&ext_str) {
        //format the result to string
        let result = format_ext_key_common(
            xprv.depth,
            xprv.parent_fingerprint,
            xprv.child_number,
            xprv.chain_code,
            &xprv.private_key.secret_bytes(), // as bytes
        );
        //return formatted string
        str_to_c_string(&result)
    } else {
        if let Ok(xpub) = Xpub::from_str(&ext_str) {
            //format the result to string
            let result = format_ext_key_common(
                xpub.depth,
                xpub.parent_fingerprint,
                xpub.child_number,
                xpub.chain_code,
                &xpub.public_key.serialize(), // as bytes
            );
            //return formatted string
            str_to_c_string(&result)
        } else {
            str_to_c_string("INVALID")
        }
    }
}
```

brunoerg commented:
Needs rebase

kuliq23 commented:
> Any update on it? It's been a month without any change here.

No update here yet. Im gonna get to it soon, hopefully next week, lots on my plate rn.

erickcestari commented:
> Glad to see a formatter! I thought it is incorporated in the CI pipeline but the fromat check seems to fail, any hints on how I avoid that?

You can run `make format-all` from the project root directory. Alternatively, run `make format` inside each module.

You can also check that the formatting is correct using the commands `make check-format-all` or `make check-format`, respectively.

The CI is displaying that the NBitcoin, basemodule, bitcoin core and rust_bitcoin modules are not correctly formatted.

erickcestari commented:
> As mentioned in https://github.com/bitcoinfuzz/bitcoinfuzz/pull/307#issuecomment-3598251317 , NBitcoin accepts keys of all 4 types that throw "non-zero child number in master key" error in rust_bitcoin and result in empty key-bytes in bitcoin-core, this makes me think that NBitcoin does not simply check the correctness as thoroughly as rust-bitcoin and bitcoin-core and this isn't a groundbreaking issue.

It's not a groundbreaking issue, but it's the type of issue that differential fuzzing is great at catching. I created a PR in NBitcoin to fix it. https://github.com/MetacoSA/NBitcoin/pull/1294.

Nice work!

erickcestari commented on modules/nbitcoin/NBitcoin/NBitcoinBridge.cs:
```diff
@@ -172,6 +173,58 @@ public static IntPtr BIP32MasterKeygen(IntPtr dataPtr, UIntPtr len)
+    {
+        if (inputPtr == IntPtr.Zero) return Marshal.StringToCoTaskMemUTF8("INVALID");
+
+        string input = Marshal.PtrToStringUTF8(inputPtr);
```
nit:

```suggestion
        string input = Marshal.PtrToStringUTF8(inputPtr) ?? "";
```

kuliq23 replied:
good idea, thanks

erickcestari commented on custommutator/mutators/extended_key.h:
```diff
@@ -0,0 +1,87 @@
+  // xprv9s21ZrQLyYpt8mej728d1sv48AxGUreFVUg2Ag6rdB8p6sHE7X7o4XxuokQTr9ucc2Wc6q5obDdTJBQFkpDTLJm7tqd1naA19inytoE1nEN
+  // std::memcpy(fuzz_data,
+  // "xprv9s21ZrQLyYpt8mej728d1sv48AxGUreFVUg2Ag6rdB8p6sHE7X7o4XxuokQTr9ucc2Wc6q5obDdTJBQFkpDTLJm7tqd1naA19inytoE1nEN",
+  // output_len);
```
nit: We can remove this hardcoded extended key comment. 
```suggestion

```

kuliq23 replied:
yes, i overlooked that, ty

erickcestari reviewed (changes requested):
Just a few nits. Then it should be ready to merge.

erickcestari reviewed (approved):
Thanks! Nice work!

ACK bd8572e05480d9306f3ad59f02ccde838c6cc061
